### PR TITLE
monotonic clock comparison

### PR DESCRIPTION
**Pull request by thillux**

### CODE REVIEW

**Comment by thillux** [comment]
Hi Stephan,

I currently don't know where to do the condition variable initialization with its attribute. Can you give me some hint how to do this or even take care of it?

BR
Markus

**Comment by smuellerDD** [comment]
> Hi Stephan,
> 
> I currently don't know where to do the condition variable initialization with its attribute. Can you give me some hint how to do this or even take care of it?

Place it next to thread_wait_cv and clear it with FINI, I would suggest.
> 
> BR Markus

**Comment by thillux** [comment]
Hi Stephan,

that should be all from my perspective. I need to do some testing now.

BR
Markus

**Comment by thillux** [comment]
I don't know if this originates in my changes, but ESDM does not close invalid client connections after timeouts.

test code:
```c++
#include <ctime>
#include <esdm/esdm_rpc_client.h>
#include <time.h>
#include <cassert>
#include <iostream>

int main(void) {
    int ret = esdm_rpcc_init_unpriv_service(nullptr);
    assert(ret == 0);

    const size_t BUF_SIZE = 32;
    uint8_t buf[BUF_SIZE];

    struct timespec ts;
    ts.tv_sec = 10;
	ts.tv_nsec = 0;

    ret = esdm_rpcc_get_random_bytes_full_timeout(buf, BUF_SIZE, &ts);
    std::cout << "return value: " << ret << std::endl;

    esdm_rpcc_fini_unpriv_service();
}
```

log messages after multiple invocations:
```
ESDM (12:47:37) (ESDM hdl_rpc7) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_seed_work_one:518]: reseed triggered by system events for DRNG on node 0
ESDM (12:47:37) (ESDM hdl_rpc7) Debug - Entropy Source [../esdm/esdm_es_aux.c:esdm_aux_get_pool:354]: obtained 0 bits by collecting 0 bits of entropy from aux pool, 0 bits of entropy remaining
ESDM (12:47:37) (ESDM hdl_rpc7) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_inject:375]: seeding regular DRNG with 48 bytes
ESDM (12:47:37) (ESDM hdl_rpc7) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_inject:388]: regular DRNG stats since last seeding: 2 secs; generate calls: 0
ESDM (12:47:45) (ESDM hdl_rpc6) Debug - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_pack_internal:196]: Server sending: server status 0, message length 11, message index 5, request ID 0
ESDM (12:47:45) (ESDM hdl_rpc6) Verbose - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_write_data:136]: Writting of data to file descriptor 6 failed: Broken pipe
ESDM (12:47:45) (ESDM hdl_rpc6) Error [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_pack_internal:210]: Submission of message data failed with error -32
ESDM (12:47:45) (ESDM hdl_rpc6) Error [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_response_closure:335]: Failed to serialize response: -32
ESDM (12:47:45) (ESDM hdl_rpc6) Debug [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_read:429]: Reading 16 bytes, already consumed 16 bytes
ESDM (12:47:45) (ESDM hdl_rpc6) Debug - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_read:448]: Server received: message length 4, message index 5, request ID 0
ESDM (12:47:45) (ESDM hdl_rpc6) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_seed_work_one:518]: reseed triggered by system events for DRNG on node 0
ESDM (12:47:45) (ESDM hdl_rpc6) Debug - Entropy Source [../esdm/esdm_es_aux.c:esdm_aux_get_pool:354]: obtained 0 bits by collecting 0 bits of entropy from aux pool, 0 bits of entropy remaining
ESDM (12:47:45) (ESDM hdl_rpc6) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_inject:375]: seeding regular DRNG with 48 bytes
ESDM (12:47:45) (ESDM hdl_rpc6) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_inject:388]: regular DRNG stats since last seeding: 8 secs; generate calls: 0
ESDM (12:47:47) (ESDM hdl_rpc7) Debug - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_pack_internal:196]: Server sending: server status 0, message length 11, message index 5, request ID 0
ESDM (12:47:47) (ESDM hdl_rpc7) Verbose - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_write_data:136]: Writting of data to file descriptor 7 failed: Broken pipe
ESDM (12:47:47) (ESDM hdl_rpc7) Error [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_pack_internal:210]: Submission of message data failed with error -32
ESDM (12:47:47) (ESDM hdl_rpc7) Error [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_response_closure:335]: Failed to serialize response: -32
ESDM (12:47:47) (ESDM hdl_rpc7) Debug [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_read:429]: Reading 16 bytes, already consumed 16 bytes
ESDM (12:47:47) (ESDM hdl_rpc7) Debug - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_read:448]: Server received: message length 4, message index 5, request ID 0
ESDM (12:47:47) (ESDM hdl_rpc7) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_seed_work_one:518]: reseed triggered by system events for DRNG on node 0
ESDM (12:47:47) (ESDM hdl_rpc7) Debug - Entropy Source [../esdm/esdm_es_aux.c:esdm_aux_get_pool:354]: obtained 0 bits by collecting 0 bits of entropy from aux pool, 0 bits of entropy remaining
ESDM (12:47:47) (ESDM hdl_rpc7) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_inject:375]: seeding regular DRNG with 48 bytes
ESDM (12:47:47) (ESDM hdl_rpc7) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_inject:388]: regular DRNG stats since last seeding: 2 secs; generate calls: 0
ESDM (12:47:55) (ESDM hdl_rpc6) Debug - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_pack_internal:196]: Server sending: server status 0, message length 11, message index 5, request ID 0
ESDM (12:47:55) (ESDM hdl_rpc6) Verbose - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_write_data:136]: Writting of data to file descriptor 6 failed: Broken pipe
ESDM (12:47:55) (ESDM hdl_rpc6) Error [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_pack_internal:210]: Submission of message data failed with error -32
ESDM (12:47:55) (ESDM hdl_rpc6) Error [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_response_closure:335]: Failed to serialize response: -32
ESDM (12:47:55) (ESDM hdl_rpc6) Debug [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_read:429]: Reading 16 bytes, already consumed 16 bytes
ESDM (12:47:55) (ESDM hdl_rpc6) Debug - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_read:448]: Server received: message length 4, message index 5, request ID 0
ESDM (12:47:55) (ESDM hdl_rpc6) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_seed_work_one:518]: reseed triggered by system events for DRNG on node 0
ESDM (12:47:55) (ESDM hdl_rpc6) Debug - Entropy Source [../esdm/esdm_es_aux.c:esdm_aux_get_pool:354]: obtained 0 bits by collecting 0 bits of entropy from aux pool, 0 bits of entropy remaining
ESDM (12:47:55) (ESDM hdl_rpc6) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_inject:375]: seeding regular DRNG with 48 bytes
ESDM (12:47:55) (ESDM hdl_rpc6) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_inject:388]: regular DRNG stats since last seeding: 8 secs; generate calls: 0
ESDM (12:47:57) (ESDM hdl_rpc7) Debug - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_pack_internal:196]: Server sending: server status 0, message length 11, message index 5, request ID 0
ESDM (12:47:57) (ESDM hdl_rpc7) Verbose - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_write_data:136]: Writting of data to file descriptor 7 failed: Broken pipe
ESDM (12:47:57) (ESDM hdl_rpc7) Error [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_pack_internal:210]: Submission of message data failed with error -32
ESDM (12:47:57) (ESDM hdl_rpc7) Error [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_response_closure:335]: Failed to serialize response: -32
ESDM (12:47:57) (ESDM hdl_rpc7) Debug [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_read:429]: Reading 16 bytes, already consumed 16 bytes
ESDM (12:47:57) (ESDM hdl_rpc7) Debug - RPC [../service-rpc/server/esdm_rpc_server.c:esdm_rpcs_read:448]: Server received: message length 4, message index 5, request ID 0
ESDM (12:47:57) (ESDM hdl_rpc7) Debug - DRNG [../esdm/esdm_drng_mgr.c:esdm_drng_seed_work_one:518]: reseed triggered by system events for DRNG on node 0
```

**Comment by smuellerDD** [comment]
Am Montag, 12. August 2024, 12:52:40 MESZ schrieb Markus Theil:

Hi Markus,

> I don't know if this originates in my changes, but ESDM does not close
> invalid client connections after timeouts.

Are you referring to the -EPIPE return code the server throws? I am not sure 
it is a real issue at this point: the thread that was spawned completes its 
operation, e.g. the DRNG reseeding, but the client already closed. Thus, the 
thread has the FD still open, wants to send the finish operation and the other 
side is already gone.

I have seen that too, when you overload slow interfaces like the _pr or the 
interface to the seed gathering. Those take a long time, but when the client 
is gone and they complete, they find a broken FD.

Bottom line, I am yet not fully sure whether this is a bug or the intended 
behavior.

Ciao
Stephan

**Comment by smuellerDD** [comment]
What do you say to this statement in the man page?

       pthread_cond_init  initializes  the  condition  variable cond, using the
       condition attributes specified in cond_attr, or  default  attributes  if
       cond_attr  is NULL.  The LinuxThreads implementation supports no attrib‐
       utes for conditions, hence the cond_attr parameter is actually ignored.


Wouldn't the patch be moot then?



Irrespective to the answer, shouldn't WAIT_QUEUE_INIT be looking more like 
this?

CKINT(pthread_mutex_init(&(name).thread_wait_lock, NULL));
CKINT(pthread_condattr_init(&attr));
ret = pthread_condattr_setclock(&attr, CLOCK_MONOTONIC);
if (!ret)
        pthread_cond_init(&(name).thread_wait_cv, &attr);
pthread_condattr_destroy(&attr);
CKINT(ret);

?

**Comment by thillux** [comment]
> What do you say to this statement in the man page?
> 
> ```
>    pthread_cond_init  initializes  the  condition  variable cond, using the
>    condition attributes specified in cond_attr, or  default  attributes  if
>    cond_attr  is NULL.  The LinuxThreads implementation supports no attrib‐
>    utes for conditions, hence the cond_attr parameter is actually ignored.
> ```
> 
> Wouldn't the patch be moot then?
>
Yes and no :) LinuxThreads was IMHO the predecessor of the current NPTL thread implementation, which supports cond attributes. See also: https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/pthread_cond_init.c;h=59b38bc59eb80eda65e7851ea2d265cad3247f21;hb=HEAD 
>
> Irrespective to the answer, shouldn't WAIT_QUEUE_INIT be looking more like this?
> 
> CKINT(pthread_mutex_init(&(name).thread_wait_lock, NULL)); CKINT(pthread_condattr_init(&attr)); ret = pthread_condattr_setclock(&attr, CLOCK_MONOTONIC); if (!ret) pthread_cond_init(&(name).thread_wait_cv, &attr); pthread_condattr_destroy(&attr); CKINT(ret);
>
I thought I don't need this, as long as the condition attribute is stack allocated. But I'll change this, as it should not harm.
> ?

**Comment by smuellerDD** [comment]
Am Montag, 12. August 2024, 15:13:17 MESZ schrieb Markus Theil:

Hi Markus,

> 
> Yes and no :) LinuxThreads was IMHO the predecessor of the current NPTL
> thread implementation, which supports cond attributes. See also:
> https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/pthread_cond_init.c;h
> =59b38bc59eb80eda65e7851ea2d265cad3247f21;hb=HEAD

Ok, thanks for the hint :-)

> > Irrespective to the answer, shouldn't WAIT_QUEUE_INIT be looking more like
> > this?
> > 
> > CKINT(pthread_mutex_init(&(name).thread_wait_lock, NULL));
> > CKINT(pthread_condattr_init(&attr)); ret =
> > pthread_condattr_setclock(&attr, CLOCK_MONOTONIC); if (!ret)
> > pthread_cond_init(&(name).thread_wait_cv, &attr);
> > pthread_condattr_destroy(&attr); CKINT(ret);
> I thought I don't need this, as long as the condition attribute is stack
> allocated. But I'll change this, as it should not harm.

I do not know the implementation of this call. But it returns ENOMEM as error 
code. This suggests to me that there is some form of memory allocation. And if 
this is true, we should release it, no?

Ciao
Stephan

**Comment by thillux** [comment]
> I do not know the implementation of this call. But it returns ENOMEM as error code. This suggests to me that there is some form of memory allocation. And if this is true, we should release it, no?

Oh, ok. I was not aware of this.

**Comment by thillux** [comment]
> Am Montag, 12. August 2024, 12:52:40 MESZ schrieb Markus Theil: Hi Markus,
> I don't know if this originates in my changes, but ESDM does not close invalid client connections after timeouts.
> Are you referring to the -EPIPE return code the server throws? I am not sure it is a real issue at this point: the thread that was spawned completes its operation, e.g. the DRNG reseeding, but the client already closed. Thus, the thread has the FD still open, wants to send the finish operation and the other side is already gone. I have seen that too, when you overload slow interfaces like the _pr or the interface to the seed gathering. Those take a long time, but when the client is gone and they complete, they find a broken FD. Bottom line, I am yet not fully sure whether this is a bug or the intended behavior. Ciao Stephan

I indeed seems to be a bug. The same fd produces multiple EPIPE errors. After the first EPIPE there is nothing left to recover IMHO. So one EPIPE for a closed connection seem to be totally fine, but not multiple.

**Comment by thillux** [comment]
Nevertheless, the EPIPE thingy is unrelated to this PR I'd say. Therefore I set it to ready now.

**Comment by thillux** [comment]
I tested the monotonic wait by setting back my system time to 2004, while waiting 10s for random bytes full. It returned after roughly 10s as designed.

**Comment by smuellerDD** [comment]
Am Montag, 12. August 2024, 15:51:31 MESZ schrieb Markus Theil:

Hi Markus,

> I tested the monotonic wait by setting back my system time to 2004, while
> waiting 10s for random bytes full. It returned after roughly 10s as
> designed.

Nice, thanks for the work.

I will check the patches tonight and will mere them then.

I will also try to look at the EPIPE issue later today.

Ciao
Stephan

**Comment by thillux** [comment]
I also managed to convert the last wait on CLOCK_REALTIME to CLOCK_MONOTONIC (at least on recent systems).

File: common/queue.h [hunk 1]
```diff
@@ -36,7 +36,11 @@ struct thread_wait_queue {
 										\
 	CKINT(pthread_mutex_init(&(name).thread_wait_lock, NULL)); 		\
 	CKINT(pthread_condattr_init(&attr));					\
-	CKINT(pthread_condattr_setclock(&attr, CLOCK_MONOTONIC));		\
+	ret = pthread_condattr_setclock(&attr, CLOCK_MONOTONIC);		\
+	if (ret) {								\
+		pthread_condattr_destroy(&attr);				\
```
Comment (smuellerDD) [inline common/queue.h]: You need to move this call down as otherwise cond_init has an empty attr.

Comment (smuellerDD) [reply common/queue.h]: Ack, jus saw it.

**Comment by thillux** [comment]
Sure, feel free to squash.

**Comment by smuellerDD** [comment]
Applied, thanks
